pubtools/_pulp/tasks/publish.py

```diff
@@ -0,0 +1,176 @@
+        # flushing caches.
+        f_sequence(publish_fs).result()
+
+        # flush UD cache
+        to_await.extend(self.flush_ud(repos))
+
+        # flush CDN cache
+        to_await.extend(self.flush_cdn(repos))
+
+        # wait for everything to finish.
+        for f in to_await:
+            f.result()
+
+        LOG.info("Publishing repositories completed")
+
+    @step("Publish")
+    def publish(self, repos):
+        out = []
+
+        publish_opts = PublishOptions(force=self.args.force, clean=self.args.clean)
+        for repo in repos:
+            LOG.info("Publishing %s", repo.id)
+            f = repo.publish(publish_opts)
+            out.append(f)
+
+        return out
+
+    @step("Flush UD cache")
+    def flush_ud(self, repos):
+        client = self.udcache_client
+        if not client:
+            LOG.info("UD cache flush is not enabled.")
+            return []
+
+        out = []
+        for repo in repos:
+            out.append(client.flush_repo(repo.id))
+
+        return out
+
+    @step("Check repos")
+    def get_repos(self):
+        repo_ids = self.args.repo_ids
+        found_repo_ids = []
+        out = []
+
+        # apply the filters and get repo_ids
+        repo_ids = self._filter_repos(repo_ids)
+
+        # get repo objects for the repo_ids
+        searched_repos = self.pulp_client.search_repository(Criteria.with_id(repo_ids))
+        for repo in searched_repos:
+            out.append(repo)
+            found_repo_ids.append(repo.id)
+
+        # Bail out if user requested repos which don't exist
+        # or there are no repos returned to publish
+        missing = set(repo_ids) - set(found_repo_ids)
+
+        missing = sorted(list(missing))
+        if missing:
+            self.fail("Requested repo(s) don't exist: %s", ", ".join(missing))
+
+        if not out:
+            self.fail("No repo(s) found to publish")
+
+        return out
+
+    def fail(self, *args, **kwargs):
+        LOG.error(*args, **kwargs)
+        sys.exit(30)
+
+    def _filter_repos(self, repos):
+
+        repos = set(repos) if repos else set()
+
+        if self.args.published_before or self.args.repo_url_regex:
+            repo_dist_f = self._filtered_repo_distributors()
+            filtered_repos = [repo_dist.repo_id for repo_dist in repo_dist_f]
+
+            return list(repos.intersection(filtered_repos)) if repos else filtered_repos
+
+        return list(repos)
+
+    def _filtered_repo_distributors(self):
+        published_before = self.args.published_before
+        url_regex = self.args.repo_url_regex
+
+        # define the criteria on available filters
+        crit = [Criteria.true()]
+        if published_before:
+            crit.append(
+                Criteria.with_field("last_publish", Matcher.less_than(published_before))
+            )
+        if url_regex:
+            crit.append(Criteria.with_field("relative_url", Matcher.regex(url_regex)))
```

Comment:
I believe this is having exactly the same bug as fixed by https://github.com/release-engineering/pubtools-pulp/pull/15

Specifically, url_regex here is a compiled regex, and that can't be serialized to JSON, so it can't be used in a real search query.  Also, a pulplib PR to detect this bug was very recently landed https://github.com/release-engineering/pubtools-pulplib/pull/68, so I'd expect tests in this repo now to fail if run against latest pulplib.

Can you please look into it and fix it if it's indeed broken?

Reply:
Yep, I retriggered the Travis build for this PR and this bug is indeed detected now by pubtools-pulplib >= 2.3.1, see https://travis-ci.org/release-engineering/pubtools-pulp/builds/592718772.

Reply:
Done